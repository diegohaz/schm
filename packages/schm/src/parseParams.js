// @flow
import { isArray, isSchema, isFunction, isObject } from './utils'
import { defaultSchema } from './schema'

const hasType = (options: any): boolean => (
  isObject(options) && options.type
)

const literalType = (options: any): Object => {
  if (isFunction(options) || isSchema(options)) {
    return { type: options }
  }
  return options
}

const defaultValue = (options: any): Object => {
  if (!isObject(options) && !isFunction(options)) {
    return { type: options.constructor, default: options }
  }
  return options
}

const nestedObject = (options: any): Object => {
  if (!isSchema(options) && isObject(options) && !hasType(options)) {
    return { type: defaultSchema(options) }
  }
  return options
}

const parseOptions = (options: any = String): Object => {
  if (isArray(options)) {
    return { type: [parseOptions(options[0])] }
  }
  if (hasType(options) && isArray(options.type)) {
    return { type: [parseOptions(options.type[0])] }
  }

  const compose = (...fns) => fns.reduce((f, g) => (...args) => f(g(...args)))
  const flow = compose(literalType, defaultValue, nestedObject)
  return flow(options)
}

const parseParams = (params: Object): Object => (
  Object.keys(params).reduce((finalParams, paramName) => ({
    ...finalParams,
    [paramName]: parseOptions(params[paramName]),
  }), {})
)

export default parseParams
