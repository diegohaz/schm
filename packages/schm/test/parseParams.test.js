import parseParams from '../src/parseParams'

it('parses literal type', () => {
  expect(parseParams({ foo: String })).toEqual({ foo: { type: String } })
})

it('parses array literal type', () => {
  expect(parseParams({ foo: [String] })).toEqual({
    foo: { type: [{ type: String }] },
  })
})

it('parses type', () => {
  expect(parseParams({ foo: { type: String } })).toEqual({
    foo: { type: String },
  })
})

it('parses array type', () => {
  expect(parseParams({ foo: [{ type: String }] })).toEqual({
    foo: { type: [{ type: String }] },
  })
})

it('parses default value', () => {
  expect(parseParams({ foo: 'bar' })).toEqual({
    foo: { type: String, default: 'bar' },
  })
})

it('parses array default value', () => {
  expect(parseParams({ foo: ['bar'] })).toEqual({
    foo: { type: [{ type: String, default: 'bar' }] },
  })
})

it('parses nested object', () => {
  expect(parseParams({ foo: { bar: String } })).toMatchSnapshot()
})

it('parses array nested object', () => {
  expect(parseParams({ foo: [{ bar: String }] })).toMatchSnapshot()
})

it('parses deep array', () => {
  expect(parseParams({ foo: [[[[[[String]]]]]] })).toMatchSnapshot()
})

it('parses type array', () => {
  expect(parseParams({ foo: { type: [{ type: String }] } })).toEqual({
    foo: { type: [{ type: String }] },
  })
})

it('parses without option', () => {
  expect(parseParams({ foo: undefined })).toEqual({ foo: { type: String } })
})
