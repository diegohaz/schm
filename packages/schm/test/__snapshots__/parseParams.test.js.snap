// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses array nested object 1`] = `
Object {
  "foo": Object {
    "type": Array [
      Object {
        "type": Object {
          "merge": [Function],
          "params": Object {
            "bar": Object {
              "type": [Function],
            },
          },
          "parse": [Function],
          "parsers": Object {
            "default": [Function],
            "get": [Function],
            "lowercase": [Function],
            "set": [Function],
            "trim": [Function],
            "type": [Function],
            "uppercase": [Function],
          },
          "validate": [Function],
          "validators": Object {
            "enum": [Function],
            "match": [Function],
            "max": [Function],
            "maxlength": [Function],
            "min": [Function],
            "minlength": [Function],
            "required": [Function],
            "type": [Function],
            "validate": [Function],
          },
        },
      },
    ],
  },
}
`;

exports[`parses deep array 1`] = `
Object {
  "foo": Object {
    "type": Array [
      Object {
        "type": Array [
          Object {
            "type": Array [
              Object {
                "type": Array [
                  Object {
                    "type": Array [
                      Object {
                        "type": Array [
                          Object {
                            "type": [Function],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
}
`;

exports[`parses nested object 1`] = `
Object {
  "foo": Object {
    "type": Object {
      "merge": [Function],
      "params": Object {
        "bar": Object {
          "type": [Function],
        },
      },
      "parse": [Function],
      "parsers": Object {
        "default": [Function],
        "get": [Function],
        "lowercase": [Function],
        "set": [Function],
        "trim": [Function],
        "type": [Function],
        "uppercase": [Function],
      },
      "validate": [Function],
      "validators": Object {
        "enum": [Function],
        "match": [Function],
        "max": [Function],
        "maxlength": [Function],
        "min": [Function],
        "minlength": [Function],
        "required": [Function],
        "type": [Function],
        "validate": [Function],
      },
    },
  },
}
`;
